### PR TITLE
feat (audience match types): Update audience evaluator and project config for new audience match types

### PR DESCRIPTION
**Pull request by yasirfolio3**
Summary

This updates project config and the audience evaluator to finish the implementation of typed audience evaluation:

Update project_config to update audience_id_map with typed_audiences
Added unit tests checking that all top-level methods that accept user attributes can bucket users into experiments or rollouts that use typed audiences.
Test Plan
New & existing unit tests
Ran compatibility suite tests for typed audiences

### CODE REVIEW

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 1]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
...
+                                                         };
+    XCTAssertTrue([self.optimizelyTypedAudience isFeatureEnabled:featureFlagKey userId:userId attributes:attributes]);
+    
+    attributes = @{
+                   @"lasers": @45.5
+                   };
+    XCTAssertTrue([self.optimizelyTypedAudience isFeatureEnabled:featureFlagKey userId:userId attributes:attributes]);
+}
+
+- (void)testIsFeatureEnabledExcludeUserFromExperimentWithTypedAudiences {
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: nit. Excludes

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 2]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
...
+    NSString *featureFlagKey = @"feat";
+    NSString *userId = @"user1";
+    NSDictionary<NSString *, NSObject *> *attributes = @{
+                                                         };
+    XCTAssertFalse([self.optimizelyTypedAudience isFeatureEnabled:featureFlagKey userId:userId attributes:attributes]);
+}
+
+- (void)testGetFeatureVariableStringReturnVariableValueWithTypedAudiences {
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: nit. Returns

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 3]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
...
+    XCTAssertEqualObjects(featureVariable, @"xyz");
+    
+    attributes = @{
+                   @"should_do_it": @true
+                   };
+    featureVariable = [self.optimizelyTypedAudience getFeatureVariableValueForType:FeatureVariableTypeString featureKey:featureKey variableKey:variableKey userId:userId attributes:attributes];
+    XCTAssertEqualObjects(featureVariable, @"xyz");
+}
+
+- (void)testGetFeatureVariableStringReturnDefaultVariableValueWithTypedAudiences {
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: nit. Returns

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 4]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
...
+    
+    OPTLYVariation *variation = [self.optimizelyTypedAudience activate:@"typed_audience_experiment" userId:@"user1" attributes:expectedAttributes callback:^(NSError *error) {
+    }];
+    XCTAssertNil(variation);
+    [expectation fulfill];
+    
+    [self waitForExpectationsWithTimeout:2 handler:nil];
+}
+
+- (void)testTrackWithTypeAudiences {
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: nit. TypedAudiences

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 5]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
...
+    
+    OPTLYVariation *variation = [self.optimizelyTypedAudience activate:@"typed_audience_experiment" userId:@"user1" attributes:expectedAttributes callback:^(NSError *error) {
+    }];
+    XCTAssertEqualObjects(@"A", variation.variationKey);
+    [expectation fulfill];
+    
+    [self waitForExpectationsWithTimeout:2 handler:nil];
+}
+
+- (void)testActivateExcludeUserFromExperimentWithTypeAudiences {
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: nit. Excludes

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 6]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
     XCTAssertEqualObjects(features, enabledFeatures);
 }
 
+#pragma mark - TypeAudiences Tests
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: Curious. Can you explain the set of tests that you have here? I see most are testing for the "happy path" i.e. everything matches and works correctly. Is that correct?

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYProjectConfigTest.m [hunk 7]
```diff
@@ -379,6 +380,25 @@ - (void)testGetAudienceForNonexistentId
     XCTAssertNil(audience, @"Shouldn't find audience for id: %@", audienceId);
 }
 
+- (void)testExperimentAudiencesRetrivedFromTypedAudiencesFirstThenFromAudiences
```
Comment (aliabbasrizvi) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYProjectConfigTest.m]: Typo. Retrieved.

**Review by aliabbasrizvi — COMMENTED** [review]
Generally looks good to me. Had a question about unit test cases.

Will leave it up to @thomaszurkan-optimizely and @jaeopt to take a closer look.

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m (hunk 6 above)
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: We have tests both for typed audience inclusion and exclusion in all APIs here. Am I missing anything?

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYProjectConfigTest.m (hunk 7 above)
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYProjectConfigTest.m]: Suggested changes made, please review.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYProjectConfig.h [hunk 8]
```diff
@@ -51,9 +51,11 @@ NS_ASSUME_NONNULL_END
 /// List of Optimizely Experiment objects
 @property (nonatomic, strong, nonnull) NSArray<OPTLYExperiment *><OPTLYExperiment> *experiments;
 /// List of Optimizely Event Type objects
-@property (nonatomic, strong, nonnull) NSArray<OPTLYEvent *><OPTLYEvent> *events;
-/// List of audience ids
-@property (nonatomic, strong, nonnull) NSArray<OPTLYAudience *><OPTLYAudience> *audiences;
+@property (nonatomic, strong, nonnull) NSArray<OPTLYEvent> *events;
```
Comment (thomaszurkan-optimizely) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYProjectConfig.h]: This is incorrect and should be reverted

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYProjectConfig.h [hunk 9]
```diff
@@ -51,9 +51,11 @@ NS_ASSUME_NONNULL_END
...
+/// List of audience objects
+@property (nonatomic, strong, nonnull) NSArray<OPTLYAudience> *audiences;
```
Comment (thomaszurkan-optimizely) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYProjectConfig.h]: Please fix these declarations to conform to JSONModel parsing (i.e @property (nonatomic, strong, nonnull) NSArray<OPTLYAudience *><OPTLYAudience> *audiences;)

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 10]
```diff
@@ -1182,6 +1193,123 @@ -(void)testGetEnabledFeaturesWithSomeFeaturesEnabledForUser {
     XCTAssertEqualObjects(features, enabledFeatures);
 }
 
+#pragma mark - TypedAudiences Tests
+
+- (void)testActivateWithTypedAudiences {
```
Comment (thomaszurkan-optimizely) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m:1198]: I'd like to see tests for active using all of the typed audience (i.e. gt, lt, eq, substring, etc.)

**Review by thomaszurkan-optimizely — CHANGES_REQUESTED** [review]
Some minor changes and a few more unit tests.

**Comment by mfahadahmed** [comment]
build

**Comment by yasirfolio3** [comment]
Base changed from yasir/audience-match-type-evaluator to yasir/audience-match-types, NAN/INF checks also implemented for audience evaluator.

**Review by thomaszurkan-optimizely — CHANGES_REQUESTED** [review]
Looks good except for the typedAudience definition.  Please fix and ping me.  I'd like to merge ASAP.

**Review by thomaszurkan-optimizely — APPROVED** [review]

**Pull request merged**